File: Top 100 Codes/04-Important Codes related to Arrays/20-Find maximum product sub-array in a given array/Findmaximumproductsubarrayinagivenarray.py
def max_product_subarray(arr):
    if not arr:
        return 0  # Edge case: empty array

    max_prod = min_prod = result = arr[0]

    for i in range(1, len(arr)):
        if arr[i] < 0:
            max_prod, min_prod = min_prod, max_prod  # Swap when encountering a negative number

        max_prod = max(arr[i], max_prod * arr[i])  # Maximum product ending at i
        min_prod = min(arr[i], min_prod * arr[i])  # Minimum product ending at i

        result = max(result, max_prod)  # Update the global max product

    return result

# Example usage
arr = [2, 3, -2, 4, -1]
print("Maximum product of subarray:", max_product_subarray(arr))

File: Top 100 Codes/04-Important Codes related to Arrays/20-Find maximum product sub-array in a given array/Findmaximumproductsubarrayinagivenarray.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

int maxProductSubarray(vector<int>& arr) {
    int n = arr.size();
    if (n == 0) 
       return 0; // Edge case

    int maxProd = arr[0], minProd = arr[0], result = arr[0];

    for (int i = 1; i < n; i++) {
        if (arr[i] < 0) 
            swap(maxProd, minProd); // Swap when encountering a negative number

        maxProd = max(arr[i], maxProd * arr[i]); // Max product ending at i
        minProd = min(arr[i], minProd * arr[i]); // Min product ending at i (handling negatives)

        result = max(result, maxProd); // Update the global max product
    }

    return result;
}

int main() {
    vector<int> arr = {2, 3, -2, 4, -1}; // Example array
    cout << "Maximum product of subarray: " << maxProductSubarray(arr) << endl;
    return 0;
}
